### PR TITLE
Allow `--resume=??` without other parameters.

### LINKED ISSUE #44: Resuming training and testing
MART implements `resume=model/checkpoint.ckpt` to resume, however, one must take care to pass `task_name` if you want it to run correctly in the appropriate directories. Similarly, one must also pass the correct `experiment=` for resume to work, even though this information is already in the resumed model's directory.

### PR DESCRIPTION
**Pull request by mzweilin**
# What does this PR do?

Fixes #44

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] `pytest tests/test_experiments.py::test_resume`

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by dxoigmn — CHANGES_REQUESTED** [review]
Can you write tests to capture these conditions? I'm forgetting now whether this solves the problem.

**Comment by mzweilin** [comment]
> Can you write tests to capture these conditions? I'm forgetting now whether this solves the problem.

@dxoigmn I added a test case and it passes. However, the new test won't capture the error in the previous code, because it exits gracefully without running an experiment.

**Review by dxoigmn — CHANGES_REQUESTED** [review]
Test looks fine but there is a logic issue in the way the resume code works. Here is where resume logic actually happens:
https://github.com/IntelLabs/MART/blob/c1064d8407e3db4d3ae8b653baabb156eede1a2b/mart/tasks/lightning.py#L30

However, because this function is wrapped by `utils.task_wrapper`:
https://github.com/IntelLabs/MART/blob/c1064d8407e3db4d3ae8b653baabb156eede1a2b/mart/tasks/lightning.py#L19-L20

Then that creates a situation where `task_wrapper` will use a cfg that is not correctly "resumed" yet. I think `utils.get_resume_checkpoint` needs to happen earlier. Can you write a test that makes sure this isn't the case?

**Comment by mzweilin** [comment]
@dxoigmn 
The test is added, and the resume logic is fixed.

File: mart/__main__.py [hunk 1]
```diff
@@ -31,19 +31,25 @@
...
         log.fatal("")
         log.fatal("Please specify an experiment to run, e.g.")
         log.fatal(
             "$ python -m mart experiment=CIFAR10_CNN fit=false +trainer.limit_test_batches=1"
         )
+        log.fatal("or specify a checkpoint to resume, e.g.")
+        log.fatal("$ python -m mart resume=logs/my_task_name/checkpoints/last.ckpt")
         log.fatal("")
-        return 0
+        raise ValueError()
```
Comment (dxoigmn) [inline mart/__main__.py:43]: I think it may be better to return 1 or -1 here instead of raising an exception.

File: tests/test_experiments.py [hunk 2]
```diff
@@ -239,3 +241,45 @@ def test_armory_carla_fasterrcnn_experiment(carla_cfg, tmp_path):
...
+                "- optimized_metric=null",  # No metric to retrieve.
+                "- extras.print_config=false",  # Test if print_config is turned off after resume.
+            ]
+        )
+    )
+
+    command = [module, "-m", "resume=" + str(ckpt)]
+    stdout_buf = run_sh_command(command)
+
+    output_dir = re.search("Output dir: (?P<path>.*?)\n", stdout_buf).group("path")
```
Comment (dxoigmn) [inline tests/test_experiments.py:284]: This feels like something you can control instead of searching for in the output?

File: tests/helpers/run_sh_command.py [hunk 3]
```diff
@@ -12,7 +14,12 @@ def run_sh_command(command: List[str]):
     """Default method for executing shell commands with pytest and sh package."""
     msg = None
     try:
-        sh.python(command, _out=lambda log: print(log.strip()))
+        with StringIO() as buf:
+            sh.python(command, _out=buf)
+            output = buf.getvalue()
+            # Remove color.
+            output = strip_ansi(output)
+        return output
```
Comment (dxoigmn) [inline tests/helpers/run_sh_command.py:22]: I think it's a bad idea to parse the output of the command. Checking exit codes is fine, but parsing output is bad. (Linux programs, in theory, shouldn't output anything with successful execution.)

Comment (dxoigmn) [reply tests/helpers/run_sh_command.py:22]: In any case, the better way to accomplish this is to tell the program to not produce colors. Does https://amoffat.github.io/sh/sections/faq.html#why-does-my-command-s-output-have-color fix this instead?

File: mart/utils/utils.py [hunk 4]
```diff
@@ -263,12 +263,12 @@ def get_resume_checkpoint(config: DictConfig) -> Tuple[DictConfig, str]:
 
         resume_checkpoint = resume_checkpoint[0]
         log.info(f"Resuming from {resume_checkpoint}")
+        # Save the ckpt_path in cfg for fit() and test().
```
Comment (dxoigmn) [inline mart/utils/utils.py:266]: This is begging for a test. It's not clear to me why `get_resume_checkpoint` has to change?

File: mart/utils/utils.py [hunk 5]
```diff
@@ -263,12 +263,12 @@ def get_resume_checkpoint(config: DictConfig) -> Tuple[DictConfig, str]:
...
+        overrides += [f"+ckpt_path={resume_checkpoint}"]
```
Comment (dxoigmn) [inline mart/utils/utils.py:267]: Why is this necessary?

**Review by dxoigmn — CHANGES_REQUESTED** [review]
More comments about logic changes.

File: tests/helpers/run_sh_command.py (hunk 3 above)
Comment (mzweilin) [reply tests/helpers/run_sh_command.py:22]: It looks like the colored module we have does not honor `_tty_out=False`.

File: tests/test_experiments.py (hunk 2 above)
Comment (mzweilin) [reply tests/test_experiments.py:284]: The output dir contains a timestamp that is out of my control.

File: mart/utils/utils.py (hunk 4 above)
Comment (mzweilin) [reply mart/utils/utils.py:266]: We need the updated check point path from `get_resume_checkpoint()` later in `fit()` or `test()`. By writing to `cfg.ckpt_path`, we avoid changing `lightning` and `task_wrapper`.

File: mart/utils/utils.py (hunk 5 above)
Comment (mzweilin) [reply mart/utils/utils.py:267]: We allow `--resume=[folder]` but `fit()` and `test()` requires a path to the `ckpt` file.

**Comment by mzweilin** [comment]
I'm separating fixes from this PR:

- #59

File: tests/helpers/run_sh_command.py (hunk 3 above)
Comment (mzweilin) [reply tests/helpers/run_sh_command.py:22]: Another option is not to use `hydra_colorlog`: `hydra/job_logging=stdout hydra/hydra_logging=hydra_debug`

File: mart/__main__.py (hunk 1 above)
Comment (mzweilin) [reply mart/__main__.py:43]: Somehow `except sh.ErrorReturnCode as e:` fails to capture the non-zero return code error. Is it because Multirun adds an extra wrapper of the program?

File: mart/utils/utils.py (hunk 4 above)
Comment (dxoigmn) [reply mart/utils/utils.py:266]: Okay I see.

File: mart/utils/utils.py (hunk 5 above)
Comment (dxoigmn) [reply mart/utils/utils.py:267]: This changes the logic of what happened before though.

Notice that before the logic enabled the user to specify `ckpt_path`. Now you override any user control. See the logic difference?

File: pyproject.toml [hunk 6]
```diff
@@ -53,7 +53,8 @@ mart = "mart.__main__:main"
 developer = [
   "pre-commit",
   "pytest",
-  "wheel"
+  "wheel",
+  "strip_ansi"
```
Comment (dxoigmn) [inline pyproject.toml:57]: I don't think we need another dependency.

File: tests/helpers/run_sh_command.py (hunk 3 above)
Comment (dxoigmn) [reply tests/helpers/run_sh_command.py:22]: What if you set `NO_COLOR=1`? See: https://github.com/borntyping/python-colorlog/blob/8390e7504612a5e44d9229b5d29dd2ca0c0f6f99/colorlog/formatter.py#LL156C30-L156C38

File: mart/__main__.py (hunk 1 above)
Comment (mzweilin) [reply mart/__main__.py:43]: I think in Python we need to run `import sys; sys.exit(-1)`.

Comment (mzweilin) [reply mart/__main__.py]: Fixed.

File: tests/helpers/run_sh_command.py (hunk 3 above)
Comment (mzweilin) [reply tests/helpers/run_sh_command.py:22]: Magic!

File: pyproject.toml (hunk 6 above)
Comment (mzweilin) [reply pyproject.toml]: Removed.

File: mart/utils/utils.py (hunk 5 above)
Comment (dxoigmn) [reply mart/utils/utils.py:267]: FYI I think there should be a test for this case.

File: tests/helpers/run_sh_command.py [hunk 7]
```diff
@@ -12,8 +13,12 @@ def run_sh_command(command: List[str]):
...
+        with StringIO() as buf:
+            sh.python(command, _out=buf, _env={"NO_COLOR": "1"})
+            output = buf.getvalue()
+        return output
     except sh.ErrorReturnCode as e:
         msg = e.stderr.decode()
-    if msg:
+    # The error message could be empty.
+    if msg is not None:
         pytest.fail(msg=msg)
```
Comment (dxoigmn) [inline tests/helpers/run_sh_command.py:24]: This is probably unnecessary for this PR now.

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -58,8 +57,7 @@ jobs:
       - name: Install dependencies
         run: |
           python -m pip install --upgrade pip
-          pip install -e .
-          pip install pytest
+          pip install -e .[developer]
           pip install pytest-cov[toml]
           pip install protobuf==3.20.0
```
Comment (dxoigmn) [inline .github/workflows/test.yml:63]: Can you move this to a different PR?

Comment (mzweilin) [reply .github/workflows/test.yml:63]: https://github.com/IntelLabs/MART/pull/59

I guess there's no `pyproject.toml` when he set up CI.

File: tests/helpers/run_sh_command.py (hunk 7 above)
Comment (mzweilin) [reply tests/helpers/run_sh_command.py:24]: I will make a separate PR.

Comment (mzweilin) [reply tests/helpers/run_sh_command.py:24]: https://github.com/IntelLabs/MART/pull/61

File: mart/utils/utils.py (hunk 5 above)
Comment (mzweilin) [reply mart/utils/utils.py:267]: I don't think it changes the program behavior. `ckpt_path` was not used because we have a more advanced `--resume` feature.

Comment (dxoigmn) [reply mart/utils/utils.py:267]: Okay if that is the case then why not just use `cfg.set("ckpt_path", resume_checkpoint)` instead of using overrides? Is there a good reason to use the overrides functionality?

**Review by dxoigmn — DISMISSED** [review]
LGTM
